src/fileAttachment.js

```diff
@@ -12,9 +13,16 @@ async function remote_fetch(file) {
 
 async function dsv(file, delimiter, {array = false, typed = false} = {}) {
   const text = await file.text();
-  return (delimiter === "\t"
-      ? (array ? tsvParseRows : tsvParse)
-      : (array ? csvParseRows : csvParse))(text, typed && autoType);
+  const parse = (delimiter === "\t"
+    ? (array ? tsvParseRows : tsvParse)
+    : (array ? csvParseRows : csvParse));
+  if (typed === "auto") {
+    const source = parse(text);
+    const schema = inferSchema(source);
+    const types = new Map(schema.map(({name, type}) => [name, type]));
+    return source.map(d => coerceRow(d, types, schema));
```

Comment:
Calling `source.map` doesn’t propagate the `source.columns` field. We could expose that here, too, but even better would be to expose the `source.schema` that we’ve already computed so that it doesn’t have to be inferred again. (Per the DatabaseClient spec, `source.schema` takes precedence over `source.columns`, so we don’t need to expose both.)

Also, it’d be nice to have a function for this operation, like say

```js
function coerceRows(source, schema) {
    const types = new Map(schema.map(({name, type}) => [name, type]));
    return source.map(d => coerceRow(d, types, schema)); // TODO expose schema
}
```

or even folding in the schema inference if necessary…

```js
function enforceSchema(source, schema = inferSchema(source)) {
    const types = new Map(schema.map(({name, type}) => [name, type]));
    return source.map(d => coerceRow(d, types, schema)); // TODO expose schema
}
```

Doesn’t really matter too much though, since it’ll presumably be only called in one place. It just might be nice since you could write nice standalone tests for it. Feels like a well-defined, self-contained operation, I mean.

Reply:
Nice 👍  Feels more readable with enforceSchema pulled out : ) and assigned the schema property to the returned array